src/main/java/com/icthh/xm/ms/entity/service/privileges/custom/EntityCustomPrivilegeService.java

```diff
@@ -0,0 +1,198 @@
+@RequiredArgsConstructor
+public class EntityCustomPrivilegeService {
+
+    private static final String CUSTOMER_PRIVILEGES_PATH = "/config/tenants/{" + TENANT_NAME + "}/custom-privileges.yml";
+
+    private final ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+
+    private final CommonConfigRepository commonConfigRepository;
+    private final PermissionProperties permissionProperties;
+    private final RoleService roleService;
+    private final List<CustomPrivilegesExtractor> privilegesExtractors;
+
+    @IgnoreLogginAspect
+    public void updateCustomPermission(Map<String, TypeSpec> specs, String tenantKey) {
+
+        String privilegesPath = resolvePathWithTenant(tenantKey, CUSTOMER_PRIVILEGES_PATH);
+        String specPath = permissionProperties.getPermissionsSpecPath();
+        String permissionsSpecPath = resolvePathWithTenant(tenantKey, specPath);
+
+        log.info("Get config from {} and {}", privilegesPath, permissionsSpecPath);
+        List<String> paths = asList(privilegesPath, permissionsSpecPath);
+        Map<String, Configuration> configs = commonConfigRepository.getConfig(null, paths);
+        configs = configs != null ? configs : new HashMap<>();
+
+        updateCustomPrivileges(specs, privilegesPath, configs.get(privilegesPath), tenantKey);
+        setNewPermissionsDefaultValue(specs, tenantKey, permissionsSpecPath, configs.get(permissionsSpecPath));
+    }
+
+    private String resolvePathWithTenant(String tenantKey, String specPath) {
+        return specPath.replace("{" + TENANT_NAME + "}", tenantKey);
+    }
+
+    @SneakyThrows
+    private void updateCustomPrivileges(Map<String, TypeSpec> specs, String privilegesPath,
+                                        Configuration customPrivileges, String tenantKey) {
+
+        val privileges = getPrivilegesConfig(customPrivileges);
+
+        addCustomPrivileges(specs, privileges, tenantKey);
+
+        String content = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(privileges);
+        if (DigestUtils.sha1Hex(content).equals(sha1Hex(customPrivileges))) {
+            log.info("Privileges configuration not changed");
+            return;
+        }
+        commonConfigRepository.updateConfigFullPath(new Configuration(privilegesPath, content), sha1Hex(customPrivileges));
+    }
+
+    private void addCustomPrivileges(Map<String, TypeSpec> specs,
+                                     Map<String, List<Map<String, Object>>> privileges,
+                                     String tenantKey) {
+        Map<String, List<Map<String, Object>>> customPrivileges = new HashMap<>();
+        for(CustomPrivilegesExtractor privilegesExtractor: privilegesExtractors) {
+            if (!privilegesExtractor.isEnabled(tenantKey)) {
+                continue;
+            }
+
+            List<Map<String, Object>> value = privilegesExtractor.toPrivileges(specs);
+            if (customPrivileges.containsKey(privilegesExtractor.getSectionName())) {
```

Comment:
Suggest rewriting if-else with `Map.merge` or `Map.compute` methods.

Reply:
in this case better to read and understand it's if else